docs(common): Add KB for partial localization/customizing selected lo… (#1437)
telerik/blazor-docs: knowledge-base/common-change-selected-localization-keys.md

---

> title: Localize selected localization keys with custom strings

Reviewer: Use title case, see [Capitalization](https://docs.telerik.com/style-guide/titles-and-headings#capitalization).
```suggestion
title: Localize Selected Localization Keys with Custom Strings
```

description: How to modify selected localized strings without maintaining the complete version of the localization resource files
type: how-to
page_title: Localize selected localization keys with custom strings
slug: common-kb-localize-selected-localization-keys
position:
tags: telerik, blazor, localization, key, resources, files, partial, string

> ticketid: 1601981, 1607465, 1604787

Reviewer 2: ```suggestion
ticketid: 1601981, 1607465, 1604787, 1608599
```

res_type: kb
---

## Environment

<table>
    <tbody>
        <tr>
            <td>Product</td>
            <td>Telerik UI for Blazor</td>
        </tr>
    </tbody>
</table>

## Description

> 

Reviewer: Are these four questions taken from four different tickets? They sound unrelated and don't build coherent text or description. This approach doesn't give significant value to the article as it represents a kind of shotgun approach where you throw all possible questions and keywords and hope that something sticks.
A much better approach is to tie them logically together and build a description that represents a single use case with multiple requirements.

If I want to change a single localized string, is there any mechanism to fall back to component integrated localization? 

Is there a way to not maintain a complete version of your localization files across all Telerik versions only because I want to change a single string? 

How to achieve partial localization? 

If I use a limited part of a component, I want to avoid providing and maintaining a full set of localized resources.

## Solution

To localize selected keys with custom strings:

> 1. As a prerequisite, create a project with Telerik UI for Blazor.

Reviewer: 1. Create a project with Telerik UI for Blazor. For detailed instructions, see:
   * [Getting Started with client-side Blazor](link)
   * [Getting Started with server-side Blazor](link)

> 2. Modify the `Program.cs` file like this:

Reviewer 2: The reader must dig to see what exactly they should add or change. Better replace **most** of the existing code with `// ...`, so that it's easier to spot the required changes. For example, see step 5.4 here - https://docs.telerik.com/blazor-ui/getting-started/server-blazor

    >caption Program.cs

>     ````CSHTML

Reviewer 2: If ResxLocalizer is in `/Services/`, then Program.cs should contain

`using TelerikBlazorPartialLocalization.Services;`

    var builder = WebApplication.CreateBuilder(args);

    // Add services to the container.
    builder.Services.AddRazorPages();
    builder.Services.AddServerSideBlazor();
    builder.Services.AddTelerikBlazor();
    // Example of how to register a service in the project (add only if such exists)
    builder.Services.AddSingleton<WeatherForecastService>();

    // register a custom localizer for the Telerik components, after registering the Telerik services
    builder.Services.AddSingleton(typeof(ITelerikStringLocalizer), typeof(ResxLocalizer));

    #region Localization

    builder.Services.AddControllers();
    builder.Services.AddLocalization(options => options.ResourcesPath = "Resources");
    builder.Services.Configure<RequestLocalizationOptions>(options =>
    {
        // define the list of cultures your app will support
        var supportedCultures = new List<CultureInfo>()
        {
            new CultureInfo("en-US")
        };

        options.SupportedCultures = supportedCultures;
        options.SupportedUICultures = supportedCultures;
    });

>     // the custom localizer service is registered later, after the Telerik services

Reviewer: ```suggestion
    // Register the custom localizer service after the Telerik services.
```

        app.UseHsts();
    }

    #region Localization

    app.UseRequestLocalization(app.Services.GetService<IOptions<RequestLocalizationOptions>>().Value);

    #endregion

    app.UseHttpsRedirection();

    app.UseStaticFiles();

    app.UseRouting();

    app.MapControllers();
    app.MapBlazorHub();
    app.MapFallbackToPage("/_Host");

    app.Run();
    ````
3. Add a cookie in the `_Host.cshtml` file to store the culture choice:
    >caption _Host.cshtml
    ````CSHTML
    @using Microsoft.AspNetCore.Localization
    @using System.Globalization

    . . . .
    <body class="k-body">

        @* Culture cookie start *@

        @{
            this.HttpContext.Response.Cookies.Append(CookieRequestCultureProvider.DefaultCookieName,
                CookieRequestCultureProvider.MakeCookieValue(new RequestCulture(CultureInfo.CurrentCulture, CultureInfo.CurrentUICulture))
            );
        }

        @* Culture cookie end *@

        <app>
            <component type="typeof(App)" render-mode="ServerPrerendered" />
        </app>

    . . . .

    </body>
    ````

> 4. Create a services folder and implement the Telerik localization service with a list of all needed localization keys and their corresponding custom strings. The service also relies on a `~/Resources` folder with the necessary `.resx` files. You can find an up-to-date list of the used strings in our [API documentation](https://docs.telerik.com/blazor-ui/api/Telerik.Blazor.Resources.Messages) and the [offline version](https://www.telerik.com/account/my-downloads) of our [Demo solution](https://demos.telerik.com/blazor-ui). Telerik maintains the main `TelerikMessages.resx` file (in English) to have the full list with each release.

Reviewer 2: It makes more sense to move step 4 before step 1. Currently the app is broken between 1 and 4, because ResxLocalizer is undefined.

    >caption ResxLocalizer.cs
    ````CSHTML
    using Telerik.Blazor.Services;

    public class ResxLocalizer : ITelerikStringLocalizer
    {
        private readonly ITelerikStringLocalizer _fallback;

        //here is the list of the desired localization keys with their custom string values
        private Dictionary<string, string> _ownStrings = new Dictionary<string, string>()
        {
            { "Wizard_Next", "Your Custom Text" }
        };

        public ResxLocalizer()
        {
            // this is your public implementation of ITelerikStringLocalizer!
            _fallback = new TelerikStringLocalizer();
        }

        public string this[string name]
        {
            get
            {
                if (_ownStrings?.ContainsKey(name) ?? false)
                {
                    return _ownStrings[name];
                }
                else
                {
                    return _fallback[name];
                }
            }
        }
    }
    ````

> 4(continued). Add the desired `.resx` files to the `~/Resources` folder.

Reviewer: Add a new step, don't use "continued".
```suggestion
5. Add the desired `.resx` files to the `~/Resources` folder.
```

>     Name the files like this `~/Resources/TelerikMessages.<culture-locale>.resx`, for example `TelerikMessages.bg-BG.resx`. You can use different names (for example, in our demos we use `TelerikMessages.resx`). The file names affect the static class that is generated and how you use it in your code (for example, to localize other elements you define yourself, such as grid command buttons or your own buttons).

Reviewer: Don't indent paragraphs, this signifies a code block in Markdown:
![image](https://github.com/telerik/blazor-docs/assets/60105689/83cfdf93-09dd-4114-8da7-a806d8898630)

>     Name the files like this `~/Resources/TelerikMessages.<culture-locale>.resx`, for example `TelerikMessages.bg-BG.resx`. You can use different names (for example, in our demos we use `TelerikMessages.resx`). The file names affect the static class that is generated and how you use it in your code (for example, to localize other elements you define yourself, such as grid command buttons or your own buttons).

Reviewer: Make the three paragraphs below a nested list (nested under step 5 above). They don't seem to follow a specific sequence, so use a bullet list.
